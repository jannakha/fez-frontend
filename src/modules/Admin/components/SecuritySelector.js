import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form/lib/immutable';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import MenuItem from '@material-ui/core/MenuItem';

import { SelectField } from 'modules/SharedComponents/Toolbox/SelectField';

import { validation } from 'config';
import { TOP_LEVEL_SECURITY_POLICIES } from 'config/general';

export const renderPolicyDesc = ({ selectedKey, policyArray = TOP_LEVEL_SECURITY_POLICIES }) => {
    if (!selectedKey) {
        return '';
    }

    const policyDesc = policyArray.find(
        policy => policy.value === selectedKey
    );

    return (
        !!policyDesc ?
            <React.Fragment>
                {policyDesc.name} ({policyDesc.id})
            </React.Fragment>
            : ''
    );
};

renderPolicyDesc.propTypes = {
    selectedKey: PropTypes.number,
    policyArray: PropTypes.array
};

export const renderPolicyItems = ({ selectedKey = null, policyArray = TOP_LEVEL_SECURITY_POLICIES }) => {
    return policyArray.map((policy, index) => {
        return (
            <MenuItem key={index} value={policy.value} selected={selectedKey === policy.value} >
                {policy.label}
            </MenuItem>
        );
    });
};

renderPolicyItems.propTypes = {
    selectedKey: PropTypes.number,
    policyArray: PropTypes.array
};

export const SecuritySelector = ({ disabled, fieldID, text, selectedPolicyKey = 0 }) => (
    <React.Fragment>
        <Grid item xs={12}>
            <Field
                component={SelectField}
                disabled={disabled}
                name={fieldID}
                label={text.fieldLabel}
                required
                validation={[validation.required]}
            >
                <MenuItem value={0} disabled>
                    {text.prompt}
                </MenuItem>
                {renderPolicyItems({
                    selectedKey: selectedPolicyKey
                })}
            </Field>
        </Grid>
        {
            selectedPolicyKey &&
            <Grid item xs={12} style={{
                padding: 24,
                backgroundColor: 'rgba(0,0,0,0.05)'
            }}>
                <Typography variant="h6" style={{ marginTop: -8 }}>
                    {text.selectedTitle}
                </Typography>
                <Grid container spacing={8} style={{ marginTop: 8 }}>
                    <Grid item xs={2}><b>Name (ID):</b></Grid>
                    <Grid item xs={10}>
                        {renderPolicyDesc({ selectedKey: selectedPolicyKey })}
                    </Grid>
                </Grid>
            </Grid>
        }
    </React.Fragment>
);

SecuritySelector.propTypes = {
    disabled: PropTypes.bool,
    selectedPolicyKey: PropTypes.number,
    text: PropTypes.object,
    fieldID: PropTypes.string,
};

export default React.memo(SecuritySelector);
