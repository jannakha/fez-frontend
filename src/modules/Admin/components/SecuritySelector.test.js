import { renderPolicyDesc, renderPolicyItems, SecuritySelector } from './SecuritySelector';

function setup(testProps, isShallow = true) {
    const props = {
        disabled: false,
        selectedPolicyKey: 1,
        fieldID: 'test1',
        text: {
            description: 'test2',
            prompt: 'test3'
        },
        ...testProps
    };

    return getElement(SecuritySelector, props, isShallow);
}

describe('SecuritySelector component', () => {
    const policyArray = [
        { name: 'test1Name', id: 'test1ID', value: 1, label: 'test1Label' },
        { name: 'test2Name', id: 'test2ID', value: 2, label: 'test2Label' },
        { name: 'test3Name', id: 'test3ID', value: 3, label: 'test3Label' },
    ];

    it('should render policy description as expected', () => {
        const test1 = renderPolicyDesc({});
        expect(test1).toMatchSnapshot();

        const test2 = renderPolicyDesc({ selectedKey: 2, policyArray });
        expect(test2).toMatchSnapshot();

        const test3 = renderPolicyDesc({ selectedKey: 9, policyArray });
        expect(test3).toMatchSnapshot();
    });

    it('should render policy items as expected', () => {
        const test1 = renderPolicyItems({});
        expect(test1).toMatchSnapshot();

        const test2 = renderPolicyItems({ policyArray });
        expect(test2).toMatchSnapshot();
    });

    it('should render properly', () => {
        const wrapper1 = setup({});
        expect(toJson(wrapper1)).toMatchSnapshot();

        const wrapper2 = setup({ selectedPolicyKey: undefined });
        expect(toJson(wrapper2)).toMatchSnapshot();
    });
});
