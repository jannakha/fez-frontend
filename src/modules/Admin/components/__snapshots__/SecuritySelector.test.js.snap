// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecuritySelector component should render policy description as expected 1`] = `""`;

exports[`SecuritySelector component should render policy description as expected 2`] = `
<React.Fragment>
  test2Name
   (
  test2ID
  )
</React.Fragment>
`;

exports[`SecuritySelector component should render policy description as expected 3`] = `""`;

exports[`SecuritySelector component should render policy items as expected 1`] = `
Array [
  <WithStyles(MenuItem)
    selected={false}
    value={1}
  >
    Administrators
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={2}
  >
    Theses Assessors
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={3}
  >
    Evidence Assessors
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={4}
  >
    Staff and Students
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={5}
  >
    Public
  </WithStyles(MenuItem)>,
]
`;

exports[`SecuritySelector component should render policy items as expected 2`] = `
Array [
  <WithStyles(MenuItem)
    selected={false}
    value={1}
  >
    test1Label
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={2}
  >
    test2Label
  </WithStyles(MenuItem)>,
  <WithStyles(MenuItem)
    selected={false}
    value={3}
  >
    test3Label
  </WithStyles(MenuItem)>,
]
`;

exports[`SecuritySelector component should render properly 1`] = `
<Fragment>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <Field
      component={[Function]}
      disabled={false}
      name="test1"
      required={true}
      validation={
        Array [
          [Function],
        ]
      }
    >
      <WithStyles(MenuItem)
        disabled={true}
        value={0}
      >
        test3
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="0"
        selected={true}
        value={1}
      >
        Administrators
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="1"
        selected={false}
        value={2}
      >
        Theses Assessors
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="2"
        selected={false}
        value={3}
      >
        Evidence Assessors
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="3"
        selected={false}
        value={4}
      >
        Staff and Students
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="4"
        selected={false}
        value={5}
      >
        Public
      </WithStyles(MenuItem)>
    </Field>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.05)",
        "padding": 24,
      }
    }
    xs={12}
  >
    <WithStyles(Typography)
      style={
        Object {
          "marginTop": -8,
        }
      }
      variant="h6"
    />
    <WithStyles(Grid)
      container={true}
      spacing={8}
      style={
        Object {
          "marginTop": 8,
        }
      }
    >
      <WithStyles(Grid)
        item={true}
        xs={2}
      >
        <b>
          Name (ID):
        </b>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={10}
      >
        Administrators
         (
        1
        )
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</Fragment>
`;

exports[`SecuritySelector component should render properly 2`] = `
<Fragment>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <Field
      component={[Function]}
      disabled={false}
      name="test1"
      required={true}
      validation={
        Array [
          [Function],
        ]
      }
    >
      <WithStyles(MenuItem)
        disabled={true}
        value={0}
      >
        test3
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="0"
        selected={false}
        value={1}
      >
        Administrators
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="1"
        selected={false}
        value={2}
      >
        Theses Assessors
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="2"
        selected={false}
        value={3}
      >
        Evidence Assessors
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="3"
        selected={false}
        value={4}
      >
        Staff and Students
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="4"
        selected={false}
        value={5}
      >
        Public
      </WithStyles(MenuItem)>
    </Field>
  </WithStyles(Grid)>
  0
</Fragment>
`;
